src/main/java/org/humancellatlas/ingest/notifications/NotificationService.java

```diff
@@ -0,0 +1,78 @@
+package org.humancellatlas.ingest.notifications;
+
+import java.time.Instant;
+import java.util.Optional;
+import java.util.stream.Stream;
+import lombok.AllArgsConstructor;
+import org.humancellatlas.ingest.notifications.exception.DuplicateNotification;
+import org.humancellatlas.ingest.notifications.model.Checksum;
+import org.humancellatlas.ingest.notifications.model.Notification;
+import org.humancellatlas.ingest.notifications.model.NotificationRequest;
+import org.humancellatlas.ingest.notifications.model.NotificationState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.dao.DuplicateKeyException;
+import org.springframework.data.rest.webmvc.ResourceNotFoundException;
+import org.springframework.stereotype.Component;
+
+
+@Component
+@AllArgsConstructor
+public class NotificationService {
+
+  private final NotificationRepository notificationRepository;
+  private final Logger log = LoggerFactory.getLogger(getClass());
+
+  public Notification createNotification(NotificationRequest notificationRequest) {
```

Comment:
They share some fields, but there isn't a duplication in terms of the intent for each. Requiring a Notification to create a Notification is paradoxical in my mind, and Notification has various attributes that should be always populated but are outside of the concern of users of this method. So the intent here is that createNotification() is solely responsible for creating Notifications given some required information (ideally, Notification's constructors would be accessible only to the NotificationService).

Reply:
> Requiring a Notification to create a Notification is paradoxical in my mind, and Notification has various attributes that should be always populated but are outside of the concern of users of this method.

If you named the method something like, `push` or `send` (i.e. `notificationService.push(notification)`) I think that would have still made sense. The act of creating `Notification`s is already being done through construction (by `NotificationBuilder`).

Reply:
The constructor doesn't really create a Notification since it doesn't persist it the entity. For this reason, I think it's better to restrict its construction in the same operation where it persisted, to avoid the possibility of null ID Notifications floating about. Ideally, I'd have its construction private to the NotificationService in an effort to "make illegal states unrepresentable".